File: app/src/main/java/com/amperas17/mangaedenapp/api/responseprovider/BaseResponseProvider.java
package com.amperas17.mangaedenapp.api.responseprovider;

import com.amperas17.mangaedenapp.utils.Caller;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public abstract class BaseResponseProvider<T, U> {

    private Caller<T> caller;
    private Call<T> call;

    public BaseResponseProvider(Caller<T> caller) {
        this.caller = caller;
    }

    public abstract Call<T> initCall(U... args);

    public void makeCall(U... args) {
        call = initCall(args);
        call.enqueue(new Callback<T>() {
            @Override
            public void onFailure(Call<T> call, Throwable t) {
                caller.onError(t);
            }

            @Override
            public void onResponse(Call<T> call, Response<T> response) {
                caller.onGetData(response.body());
            }
        });
    }

    public void stopRequest() {
        if (call != null && call.isExecuted())
            call.cancel();
    }
}

File: app/src/main/java/com/amperas17/mangaedenapp/api/responseprovider/BaseResponseProvider.kt
package com.amperas17.mangaedenapp.api.responseprovider

import com.amperas17.mangaedenapp.api.MangaApi
import com.amperas17.mangaedenapp.utils.Caller
import org.koin.standalone.KoinComponent
import org.koin.standalone.inject

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


abstract class BaseResponseProvider<T, U>(private val caller: Caller<T>): KoinComponent {
    protected val mangaApi by inject<MangaApi>()
    private var call: Call<T>? = null

    abstract fun initCall(vararg args: U): Call<T>

    fun makeCall(vararg args: U) {
        call = initCall(*args)
        call?.enqueue(object : Callback<T> {
            override fun onFailure(call: Call<T>, t: Throwable) {
                caller.onError(t)
            }

            override fun onResponse(call: Call<T>, response: Response<T>) {
                response.body()?.let { caller.onGetData(it) }
            }
        })
    }

    fun stopRequest() {
        if (call?.isExecuted == true) {
            call?.cancel()
        }
    }
}
